Add form, presenter

andersodt opened the pull request.

andersodt committed:
Add form, presenter

kkestell commented on lib/fern/api.rb:
```diff
@@ -8,14 +8,43 @@ module Api
+      def form
+        if @form.nil?
+          p = params
+          if fern[action_name.to_sym][:form].has_key?(:key) && fern[action_name.to_sym][:form][:key]
```
Might be worthwhile to assign `fern[action_name.to_sym][:form]` to a variable? It's used twice on this line, again on line 16, and again on line 19.

kkestell commented on lib/fern/api.rb:
```diff
@@ -8,14 +8,43 @@ module Api
+
+      def form
+        if @form.nil?
+          p = params
```
Do we want to support strong params here, I wonder? Cerulean included the following:
```ruby
p = p.permit! if params.respond_to?(:permit!)
```

kkestell commented on lib/fern/api.rb:
```diff
@@ -8,14 +8,43 @@ module Api
+        render json: render_presenter(object, **opts)
+      end
+
+      def render_presenter(object, **opts)
```
I'm a little torn about this method. It's certainly easy to grok (although we may consider drying up `fern[action_name.to_sym][:presenter]` which appears in here four times), but lines 31 and 33, and 37 and 39 are so similar, I can't help but think there's a more concise way to express the logic. Any ideas?

andersodt replied:
```         
 if opts.any?
            fern[action_name.to_sym][:presenter].new(object).as_json(**opts)
          else
            fern[action_name.to_sym][:presenter].new(object).as_json()
          end
```

is it possible to simply just pass `**opts` in any case? I see this around much of the presenter logic

andersodt replied:
```      
      def render_presenter(object, **opts)
        presenter = fern[action_name.to_sym][:presenter]

        if object.respond_to?(:map)
          object.map { |el| presenter.new(el).as_json(**opts) }
        else
          presenter.new(object).as_json(**opts)
        end
      end
```

Was the original intent to not pass around empty hashes if not necessary?

kkestell replied:
Hmm. I wish I could remember! I suspect I was trying to prevent you from always having to declare your `as_json` method as taking an argument. Otherwise you'd get a `ArgumentError: wrong number of arguments (given 1, expected 0)`.

kkestell commented on lib/fern/api.rb:
```diff
@@ -8,14 +8,43 @@ module Api
+      end
+
+      def render_presenter(object, **opts)
+        if object.is_a?(Array) || object.is_a?(ActiveRecord::Relation)
```
I wonder if there's a better way to do this. Maybe `respond_to?(:map)`? What we really care about here is whether or not `object` quacks like a list.

andersodt replied:
I'm not sure what this is about. It looks to be supporting something like `present multiple_items`, but the only time we really present multiple objects is with a `SearchResultsPresenter`, which actually doesn't trigger this conditional because `present results...` is a single `Hash`, not an Array or Relation.

In the `SearchResults...` case, your suggestion wouldn't work because `Hash` does respond to map (but we don't want to call a new presenter for each item in the Hash).

kkestell commented on lib/fern/api.rb:
```diff
@@ -25,4 +54,4 @@ def endpoint(name, &block)
 end
 
-ActiveSupport.on_load(:action_controller) { include Fern::Api }
+# ActiveSupport.on_load(:action_controller) { include Fern::Api }
```
Bummer this didn't work; I sorta liked it. We should probably just remove it instead of commenting it out though.

andersodt replied:
I think it _could_ work, with some thought -- the issue is that when you have other controllers that inherit `ActionController` that do work on params `params` (like Devise). 

I don't know much about `on_load` -- I was thinking maybe, instead of having to define `FernController` on the project level, have a base `FernController` class defined in this gem?

andersodt replied:
I definitely support the idea of doing as little work as possible to include this in your project

kkestell commented:
Oh! Like a dullard, I broke CI with my last commit. I wonder what needs to be done to make the tests pass again? Ideally we'd have tests for these new methods. If it's not a ton of work, maybe you want to fix whatever's broken with the build while you're in there?

andersodt committed:
Clean up some code

andersodt committed:
Improve form, get tests to pass

andersodt committed:
Removing trailing whitespace

andersodt committed:
Fix expand_paths

andersodt commented:
Ok -- I rewrote some of the form and presenter logic and added a comment about one of those statements. I'm thinking that the next step would be to completely remove all of the form stuff from this repo and move it all to a new `fern-form`, similar to `fern-parameters`. Thoughts? Ideally, I wouldn't merge this PR until the form stuff has been broken out.

andersodt committed:
Remove Form from API Root

andersodt committed:
Merge branch 'master' into origin-updates

andersodt committed:
Fix Rubocop Violation
